tests/aead.cc

```diff
@@ -0,0 +1,183 @@
+
+// ----- EverCrypt -------------------------------------------------------------
+
+void
+encrypt_decrypt(EverCrypt_AEAD_state_s* state,
+                bytes& iv,
+                bytes& aad,
+                bytes& msg,
+                bytes& cipher,
+                bytes& tag,
+                bool valid)
+{
+  // Encrypt
+  {
+    bytes got_cipher(cipher.size());
+    bytes got_tag(tag.size());
+    EverCrypt_Error_error_code res = EverCrypt_AEAD_encrypt(state,
+                                                            iv.data(),
+                                                            iv.size(),
+                                                            aad.data(),
+                                                            aad.size(),
+                                                            msg.data(),
+                                                            msg.size(),
+                                                            got_cipher.data(),
+                                                            got_tag.data());
+
+    if (valid) {
+      ASSERT_EQ(res, 0);
+      ASSERT_EQ(cipher, got_cipher);
+      ASSERT_EQ(tag, got_tag);
+    }
+  }
+
+  // Decrypt
+  {
+    bytes got_msg(msg.size());
+    EverCrypt_Error_error_code res = EverCrypt_AEAD_decrypt(state,
+                                                            iv.data(),
+                                                            iv.size(),
+                                                            aad.data(),
+                                                            aad.size(),
+                                                            cipher.data(),
+                                                            cipher.size(),
+                                                            tag.data(),
+                                                            got_msg.data());
+
+    if (valid) {
+      ASSERT_EQ(res, 0);
+      ASSERT_EQ(msg, got_msg);
+    } else {
+      ASSERT_NE(res, 0);
+    }
+  }
+}
+
+// ----- ChaCha20Poly1305 ------------------------------------------------------
+
+typedef EverCryptSuite<WycheproofAeadTest> ChaCha20Suite;
+
+TEST_P(ChaCha20Suite, KAT)
+{
+  EverCryptConfig config;
+  WycheproofAeadTest test;
+  tie(config, test) = this->GetParam();
+
+  EverCrypt_AEAD_state_s* state;
+  EverCrypt_Error_error_code res = EverCrypt_AEAD_create_in(
+    Spec_Agile_AEAD_CHACHA20_POLY1305, &state, test.key.data());
+  // Should always work.
+  ASSERT_EQ(res, EverCrypt_Error_Success);
+
+  encrypt_decrypt(
+    state, test.iv, test.aad, test.msg, test.ct, test.tag, test.valid);
+
+  EverCrypt_AEAD_free(state);
+}
+
+// ----- AES GCM -------------------------------------------------------------
+
+typedef EverCryptSuite<WycheproofAeadTest> AesGcmSuite;
+
+TEST_P(AesGcmSuite, KAT)
+{
+  EverCryptConfig config;
+  WycheproofAeadTest test;
+  tie(config, test) = this->GetParam();
+
+  EverCrypt_AEAD_state_s* state;
```

Comment:
Looks like this is never freed?

Reply:
[0e61c4b](https://github.com/cryspen/hacl-packages/pull/145/commits/0e61c4b0e75c8e702562ab8a70dddb979095899d)